src/utils/stream.ts

```diff
@@ -0,0 +1,399 @@
+import { Address as BitcoreAddress, HDPublicKey } from 'bitcore-lib';
+import FullnodeConnection from '../new/connection';
+import { IStorage, IHistoryTx, HistorySyncMode, isGapLimitScanPolicy } from '../types';
+import Network from '../models/network';
+
+interface IStreamSyncHistoryVertex {
+  type: 'stream:history:vertex';
+  id: string;
+  data: IHistoryTx;
+}
+
+interface IStreamSyncHistoryAddress {
+  type: 'stream:history:address';
+  id: string;
+  address: string;
+  index: number;
+}
+
+interface IStreamSyncHistoryEnd {
+  type: 'stream:history:end';
+  id: string;
+}
+
+interface IStreamSyncHistoryError {
+  type: 'stream:history:error';
+  id: string;
+  errmsg: string;
+}
+
+type IStreamSyncHistoryData =
+  | IStreamSyncHistoryVertex
+  | IStreamSyncHistoryAddress
+  | IStreamSyncHistoryEnd
+  | IStreamSyncHistoryError;
+
+function isStreamSyncHistoryVertex(data: IStreamSyncHistoryData): data is IStreamSyncHistoryVertex {
+  return data.type === 'stream:history:vertex';
+}
+
+function isStreamSyncHistoryAddress(
+  data: IStreamSyncHistoryData
+): data is IStreamSyncHistoryAddress {
+  return data.type === 'stream:history:address';
+}
+
+function isStreamSyncHistoryEnd(data: IStreamSyncHistoryData): data is IStreamSyncHistoryEnd {
+  return data.type === 'stream:history:end';
+}
+
+function isStreamSyncHistoryError(data: IStreamSyncHistoryData): data is IStreamSyncHistoryError {
+  return data.type === 'stream:history:error';
+}
+
+/**
+ * Load addresses in a CPU intensive way
+ * This only contemplates P2PKH addresses for now.
+ */
+export function loadAddressesCPUIntensive(
+  startIndex: number,
+  count: number,
+  xpubkey: string,
+  networkName: string
+): [number, string][] {
+  const addresses: [number, string][] = [];
+  const stopIndex = startIndex + count;
+  const network = new Network(networkName);
+  const hdpubkey = new HDPublicKey(xpubkey);
+
+  for (let i = startIndex; i < stopIndex; i++) {
+    const key = hdpubkey.deriveChild(i);
+    addresses.push([i, new BitcoreAddress(key.publicKey, network.bitcoreNetwork).toString()]);
+  }
+
+  return addresses;
+}
+
+export function generateStreamId() {
+  return Math.random().toString(36).substring(2, 15);
+}
+
+export async function xpubStreamSyncHistory(
+  startIndex: number,
+  _count: number,
+  storage: IStorage,
+  connection: FullnodeConnection,
+  shouldProcessHistory: boolean = false
+) {
+  let firstIndex = startIndex;
+  const scanPolicyData = await storage.getScanningPolicyData();
+  if (isGapLimitScanPolicy(scanPolicyData)) {
```

Comment:
What if it is not? What are the other policies? Should we support them too?

Reply:
I agree. Let's add it as a task for the next phase of this project. Then we can prioritize it later.

Reply:
You can close this thread after you add a new task to get this done later in the project.

Reply:
The issue https://github.com/HathorNetwork/hathor-wallet-lib/issues/709 was created to track this.